include/quic/endpoint.hpp

```diff
@@ -20,52 +20,130 @@ extern "C"
 #include "utils.hpp"
 
 namespace oxen::quic
 {
-    class Connection;
-    class Handler;
+	using conn_ptr_pair = std::pair<std::shared_ptr<Connection>, std::shared_ptr<connection_interface>>;
 
-    class Endpoint
+    class Endpoint : std::enable_shared_from_this<Endpoint>
     {
+		friend class Network;
         friend class Connection;
+		friend class Stream;
 
-      public:
-        explicit Endpoint(std::shared_ptr<Handler>& quic_manager);
-        virtual ~Endpoint() { log::trace(log_cat, "{} called", __PRETTY_FUNCTION__); }
+      private:
+		const Address local;
+        std::shared_ptr<uvw::timer_handle> expiry_timer;
+        std::shared_ptr<uv_udp_t> handle;
+		config_t user_config{};
+		bool accepting_inbound{false};
+        Network& net;
 
-        std::shared_ptr<Handler> handler;
+      public:
+        explicit Endpoint(Network& n, const Address& listen_addr);
+
+		template <typename... Opt>
+		bool inbound_init(Opt&&... opts)
+		{
+			std::promise<bool> p;
+			auto f = p.get_future();
+
+			net.call([&opts..., &p, this]() mutable {
+				try
+				{
+					// initialize client context and client tls context simultaneously
+					inbound_ctx = std::make_shared<InboundContext>(std::forward<Opt>(opts)...);
+					accepting_inbound = true;
+
+					log::debug(log_cat, "Inbound context ready for incoming connections");
+
+					p.set_value(true);
+				}
+				catch (...)
+				{
+					p.set_exception(std::current_exception());
+				}
+			});
+
+			return f.get();
+		}
+
+
+        // creates new outbound connection to remote; emplaces conn/interface pair in outbound map
+		template <typename... Opt>
+        std::shared_ptr<connection_interface> connect(const Address& remote, Opt&&... opts)
+		{
+			std::promise<std::shared_ptr<connection_interface>> p;
+			auto f = p.get_future();
+
+			net.call([&opts..., &p, this, raddr = remote]() mutable {
+				try
+				{
+					// initialize client context and client tls context simultaneously
+					outbound_ctx = std::make_shared<OutboundContext>(std::forward<Opt>(opts)...);
+
+					if (auto [itr, res] = outbound_conns.emplace(std::piecewise_construct,
+						std::forward_as_tuple(ConnectionID::random()), std::forward_as_tuple(nullptr, nullptr)); res)
+					{
+						itr->second = std::move(Connection::make_outbound_conn_pair(
+							*this,
+							itr->first,
+							ConnectionID::random(),
+							Path{local, raddr},
+							handle,
+							outbound_ctx->tls_creds,
+							user_config
+						));
+
+						p.set_value(itr->second.second);
+					}
+					else
+						p.set_value(nullptr);
+				}
+				catch (...)
+				{
+					p.set_exception(std::current_exception());
+				}
+			});
+
+			return f.get();
+		};
 
         std::shared_ptr<uvw::loop> get_loop();
 
-        void handle_packet(Packet& pkt);
-
-        void close_connection(Connection& conn, int code = NGTCP2_NO_ERROR, std::string_view msg = "NO_ERROR"sv);
+        Connection* get_conn_ptr(ConnectionID ID);		// query by conn ID
```

Comment:
This and the one below should be private so library users don't think they should be calling them.